The job **`aarch64-gnu-llvm-19-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/17300230287/job/49109197170) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/49109197170) [(plain)](https://github.com/rust-lang/rust/commit/8bd5fb7e5a9371cc0be2b14e7de7185e41348d36/checks/49109197170/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
---- [ui] tests/ui/force-inlining/cross-crate.rs stdout ----

error: auxiliary build of /checkout/tests/ui/force-inlining/auxiliary/callees.rs failed to compile: 
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/force-inlining/auxiliary/callees.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--target=aarch64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/force-inlining/cross-crate/auxiliary" "-A" "internal_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers" "--crate-type=lib" "--crate-type" "dylib" "-L" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/force-inlining/cross-crate/auxiliary"
stdout: none
--- stderr -------------------------------
error: linking with `cc` failed: exit status: 1
   |
   = note:  "cc" "-Wl,--version-script=/tmp/rustcVp6t40/list" "-Wl,--no-undefined-version" "/tmp/rustcVp6t40/symbols.o" "<1 object files omitted>" "/tmp/rustcVp6t40/rmeta.o" "-Wl,--as-needed" "-Wl,-Bdynamic" "<sysroot>/lib/rustlib/aarch64-unknown-linux-gnu/lib/libstd-9ffeec687794d6ca.so" "-Wl,-Bstatic" "<sysroot>/lib/rustlib/aarch64-unknown-linux-gnu/lib/libcompiler_builtins-*.rlib" "-Wl,-Bdynamic" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-L" "/tmp/rustcVp6t40/raw-dylibs" "-Wl,--eh-frame-hdr" "-Wl,-z,noexecstack" "-L" "/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/force-inlining/cross-crate/auxiliary" "-L" "<sysroot>/lib/rustlib/aarch64-unknown-linux-gnu/lib" "-o" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/force-inlining/cross-crate/auxiliary/libcallees.so" "-shared" "-Wl,-soname=libcallees.so" "-Wl,-z,relro,-z,now" "-Wl,--strip-debug" "-nodefaultlibs" "-Wl,-rpath,$ORIGIN/../../../../../stage2/lib/rustlib/aarch64-unknown-linux-gnu/lib,--enable-new-dtags,-z,origin"
   = note: some arguments are omitted. use `--verbose` to show all linker arguments
   = note: /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc42___rust_alloc_error_handler_should_panic_v2: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc35___rust_no_alloc_shim_is_unstable_v2: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc26___rust_alloc_error_handler: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc19___rust_alloc_zeroed: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc14___rust_realloc: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc14___rust_dealloc: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc12___rust_alloc: undefined version: 
           /usr/bin/ld: failed to set dynamic section sizes: bad value
           collect2: error: ld returned 1 exit status
           

error: aborting due to 1 previous error
------------------------------------------


---- [ui] tests/ui/lto/debuginfo-lto.rs stdout ----

error: auxiliary build of /checkout/tests/ui/lto/auxiliary/debuginfo-lto-aux.rs failed to compile: 
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/lto/auxiliary/debuginfo-lto-aux.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--target=aarch64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/lto/debuginfo-lto/auxiliary" "-A" "internal_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Lnative=/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers" "-g" "--crate-type=rlib" "--crate-type" "dylib" "-L" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/lto/debuginfo-lto/auxiliary"
stdout: none
--- stderr -------------------------------
warning: field `0` is never read
##[warning]  --> /checkout/tests/ui/lto/auxiliary/debuginfo-lto-aux.rs:3:35
   |
---
   = help: consider removing this field

error: linking with `cc` failed: exit status: 1
   |
   = note:  "cc" "-Wl,--version-script=/tmp/rustc6lNC65/list" "-Wl,--no-undefined-version" "/tmp/rustc6lNC65/symbols.o" "<1 object files omitted>" "/tmp/rustc6lNC65/rmeta.o" "-Wl,--as-needed" "-Wl,-Bdynamic" "<sysroot>/lib/rustlib/aarch64-unknown-linux-gnu/lib/libstd-9ffeec687794d6ca.so" "-Wl,-Bstatic" "<sysroot>/lib/rustlib/aarch64-unknown-linux-gnu/lib/libcompiler_builtins-*.rlib" "-Wl,-Bdynamic" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-L" "/tmp/rustc6lNC65/raw-dylibs" "-Wl,--eh-frame-hdr" "-Wl,-z,noexecstack" "-L" "/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/lto/debuginfo-lto/auxiliary" "-L" "<sysroot>/lib/rustlib/aarch64-unknown-linux-gnu/lib" "-o" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/lto/debuginfo-lto/auxiliary/libdebuginfo_lto_aux.so" "-shared" "-Wl,-soname=libdebuginfo_lto_aux.so" "-Wl,-z,relro,-z,now" "-Wl,--strip-debug" "-nodefaultlibs" "-Wl,-rpath,$ORIGIN/../../../../../stage2/lib/rustlib/aarch64-unknown-linux-gnu/lib,--enable-new-dtags,-z,origin"
   = note: some arguments are omitted. use `--verbose` to show all linker arguments
   = note: /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc42___rust_alloc_error_handler_should_panic_v2: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc35___rust_no_alloc_shim_is_unstable_v2: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc26___rust_alloc_error_handler: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc19___rust_alloc_zeroed: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc14___rust_realloc: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc14___rust_dealloc: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc12___rust_alloc: undefined version: 
           /usr/bin/ld: failed to set dynamic section sizes: bad value
           collect2: error: ld returned 1 exit status
           

error: aborting due to 1 previous error; 2 warnings emitted
------------------------------------------


---- [ui] tests/ui/lto/dwarf-mixed-versions-lto.rs stdout ----

error: auxiliary build of /checkout/tests/ui/lto/auxiliary/dwarf-mixed-versions-lto-aux.rs failed to compile: 
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/lto/auxiliary/dwarf-mixed-versions-lto-aux.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--target=aarch64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/lto/dwarf-mixed-versions-lto/auxiliary" "-A" "internal_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Lnative=/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers" "-g" "--crate-type=rlib" "-Cdwarf-version=4" "--crate-type" "dylib" "-L" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/lto/dwarf-mixed-versions-lto/auxiliary"
stdout: none
--- stderr -------------------------------
error: linking with `cc` failed: exit status: 1
   |
   = note:  "cc" "-Wl,--version-script=/tmp/rustcAEQSJl/list" "-Wl,--no-undefined-version" "/tmp/rustcAEQSJl/symbols.o" "<1 object files omitted>" "/tmp/rustcAEQSJl/rmeta.o" "-Wl,--as-needed" "-Wl,-Bdynamic" "<sysroot>/lib/rustlib/aarch64-unknown-linux-gnu/lib/libstd-9ffeec687794d6ca.so" "-Wl,-Bstatic" "<sysroot>/lib/rustlib/aarch64-unknown-linux-gnu/lib/libcompiler_builtins-*.rlib" "-Wl,-Bdynamic" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-L" "/tmp/rustcAEQSJl/raw-dylibs" "-Wl,--eh-frame-hdr" "-Wl,-z,noexecstack" "-L" "/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/lto/dwarf-mixed-versions-lto/auxiliary" "-L" "<sysroot>/lib/rustlib/aarch64-unknown-linux-gnu/lib" "-o" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/lto/dwarf-mixed-versions-lto/auxiliary/libdwarf_mixed_versions_lto_aux.so" "-shared" "-Wl,-soname=libdwarf_mixed_versions_lto_aux.so" "-Wl,-z,relro,-z,now" "-Wl,--strip-debug" "-nodefaultlibs" "-Wl,-rpath,$ORIGIN/../../../../../stage2/lib/rustlib/aarch64-unknown-linux-gnu/lib,--enable-new-dtags,-z,origin"
   = note: some arguments are omitted. use `--verbose` to show all linker arguments
   = note: /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc42___rust_alloc_error_handler_should_panic_v2: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc35___rust_no_alloc_shim_is_unstable_v2: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc26___rust_alloc_error_handler: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc19___rust_alloc_zeroed: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc14___rust_realloc: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc14___rust_dealloc: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc12___rust_alloc: undefined version: 
           /usr/bin/ld: failed to set dynamic section sizes: bad value
           collect2: error: ld returned 1 exit status
           

error: aborting due to 1 previous error
------------------------------------------


---- [ui] tests/ui/lto/msvc-imp-present.rs stdout ----

error: auxiliary build of /checkout/tests/ui/lto/auxiliary/msvc-imp-present.rs failed to compile: 
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/lto/auxiliary/msvc-imp-present.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--target=aarch64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--out-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/lto/msvc-imp-present/auxiliary" "-A" "internal_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers" "-Z" "thinlto" "-C" "codegen-units=8" "-C" "prefer-dynamic" "-L" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/lto/msvc-imp-present/auxiliary"
stdout: none
--- stderr -------------------------------
error: linking with `cc` failed: exit status: 1
   |
   = note:  "cc" "-Wl,--version-script=/tmp/rustc4uueVz/list" "-Wl,--no-undefined-version" "/tmp/rustc4uueVz/symbols.o" "<2 object files omitted>" "/tmp/rustc4uueVz/rmeta.o" "-Wl,--as-needed" "-Wl,-Bdynamic" "<sysroot>/lib/rustlib/aarch64-unknown-linux-gnu/lib/libstd-9ffeec687794d6ca.so" "-Wl,-Bstatic" "<sysroot>/lib/rustlib/aarch64-unknown-linux-gnu/lib/libcompiler_builtins-*.rlib" "-Wl,-Bdynamic" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-L" "/tmp/rustc4uueVz/raw-dylibs" "-Wl,--eh-frame-hdr" "-Wl,-z,noexecstack" "-L" "/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/lto/msvc-imp-present/auxiliary" "-L" "<sysroot>/lib/rustlib/aarch64-unknown-linux-gnu/lib" "-o" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/lto/msvc-imp-present/auxiliary/libmsvc_imp_present.so" "-shared" "-Wl,-soname=libmsvc_imp_present.so" "-Wl,-z,relro,-z,now" "-Wl,--strip-debug" "-nodefaultlibs" "-Wl,-rpath,$ORIGIN/../../../../../stage2/lib/rustlib/aarch64-unknown-linux-gnu/lib,--enable-new-dtags,-z,origin"
   = note: some arguments are omitted. use `--verbose` to show all linker arguments
   = note: /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc42___rust_alloc_error_handler_should_panic_v2: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc35___rust_no_alloc_shim_is_unstable_v2: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc26___rust_alloc_error_handler: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc19___rust_alloc_zeroed: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc14___rust_realloc: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc14___rust_dealloc: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc12___rust_alloc: undefined version: 
           /usr/bin/ld: failed to set dynamic section sizes: bad value
           collect2: error: ld returned 1 exit status
           

error: aborting due to 1 previous error
------------------------------------------


---- [ui] tests/ui/sepcomp/sepcomp-lib-lto.rs stdout ----

error: auxiliary build of /checkout/tests/ui/sepcomp/auxiliary/sepcomp_lib.rs failed to compile: 
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/sepcomp/auxiliary/sepcomp_lib.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--target=aarch64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/sepcomp/sepcomp-lib-lto/auxiliary" "-A" "internal_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Lnative=/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers" "-C" "codegen-units=3" "--crate-type=rlib,dylib" "-g" "--crate-type" "dylib" "-L" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/sepcomp/sepcomp-lib-lto/auxiliary"
stdout: none
--- stderr -------------------------------
error: linking with `cc` failed: exit status: 1
   |
   = note:  "cc" "-Wl,--version-script=/tmp/rustcCGPPxe/list" "-Wl,--no-undefined-version" "/tmp/rustcCGPPxe/symbols.o" "<3 object files omitted>" "/tmp/rustcCGPPxe/rmeta.o" "-Wl,--as-needed" "-Wl,-Bdynamic" "<sysroot>/lib/rustlib/aarch64-unknown-linux-gnu/lib/libstd-9ffeec687794d6ca.so" "-Wl,-Bstatic" "<sysroot>/lib/rustlib/aarch64-unknown-linux-gnu/lib/libcompiler_builtins-*.rlib" "-Wl,-Bdynamic" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-L" "/tmp/rustcCGPPxe/raw-dylibs" "-Wl,--eh-frame-hdr" "-Wl,-z,noexecstack" "-L" "/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/sepcomp/sepcomp-lib-lto/auxiliary" "-L" "<sysroot>/lib/rustlib/aarch64-unknown-linux-gnu/lib" "-o" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/sepcomp/sepcomp-lib-lto/auxiliary/libsepcomp_lib.so" "-shared" "-Wl,-soname=libsepcomp_lib.so" "-Wl,-z,relro,-z,now" "-Wl,--strip-debug" "-nodefaultlibs" "-Wl,-rpath,$ORIGIN/../../../../../stage2/lib/rustlib/aarch64-unknown-linux-gnu/lib,--enable-new-dtags,-z,origin"
   = note: some arguments are omitted. use `--verbose` to show all linker arguments
   = note: /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc42___rust_alloc_error_handler_should_panic_v2: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc35___rust_no_alloc_shim_is_unstable_v2: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc26___rust_alloc_error_handler: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc19___rust_alloc_zeroed: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc14___rust_realloc: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc14___rust_dealloc: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc12___rust_alloc: undefined version: 
           /usr/bin/ld: failed to set dynamic section sizes: bad value
           collect2: error: ld returned 1 exit status
           

error: aborting due to 1 previous error
------------------------------------------


---- [ui] tests/ui/sepcomp/sepcomp-lib.rs stdout ----

error: auxiliary build of /checkout/tests/ui/sepcomp/auxiliary/sepcomp_lib.rs failed to compile: 
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/sepcomp/auxiliary/sepcomp_lib.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--target=aarch64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/sepcomp/sepcomp-lib/auxiliary" "-A" "internal_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Lnative=/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers" "-C" "codegen-units=3" "--crate-type=rlib,dylib" "-g" "--crate-type" "dylib" "-L" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/sepcomp/sepcomp-lib/auxiliary"
stdout: none
--- stderr -------------------------------
error: linking with `cc` failed: exit status: 1
   |
   = note:  "cc" "-Wl,--version-script=/tmp/rustcjruR9U/list" "-Wl,--no-undefined-version" "/tmp/rustcjruR9U/symbols.o" "<3 object files omitted>" "/tmp/rustcjruR9U/rmeta.o" "-Wl,--as-needed" "-Wl,-Bdynamic" "<sysroot>/lib/rustlib/aarch64-unknown-linux-gnu/lib/libstd-9ffeec687794d6ca.so" "-Wl,-Bstatic" "<sysroot>/lib/rustlib/aarch64-unknown-linux-gnu/lib/libcompiler_builtins-*.rlib" "-Wl,-Bdynamic" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-L" "/tmp/rustcjruR9U/raw-dylibs" "-Wl,--eh-frame-hdr" "-Wl,-z,noexecstack" "-L" "/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/sepcomp/sepcomp-lib/auxiliary" "-L" "<sysroot>/lib/rustlib/aarch64-unknown-linux-gnu/lib" "-o" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/sepcomp/sepcomp-lib/auxiliary/libsepcomp_lib.so" "-shared" "-Wl,-soname=libsepcomp_lib.so" "-Wl,-z,relro,-z,now" "-Wl,--strip-debug" "-nodefaultlibs" "-Wl,-rpath,$ORIGIN/../../../../../stage2/lib/rustlib/aarch64-unknown-linux-gnu/lib,--enable-new-dtags,-z,origin"
   = note: some arguments are omitted. use `--verbose` to show all linker arguments
   = note: /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc42___rust_alloc_error_handler_should_panic_v2: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc35___rust_no_alloc_shim_is_unstable_v2: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc26___rust_alloc_error_handler: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc19___rust_alloc_zeroed: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc14___rust_realloc: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc14___rust_dealloc: undefined version: 
           /usr/bin/ld: _RNvCs6qA6grgM2AZ_7___rustc12___rust_alloc: undefined version: 
           /usr/bin/ld: failed to set dynamic section sizes: bad value
           collect2: error: ld returned 1 exit status
           

error: aborting due to 1 previous error
------------------------------------------
```

</details>

--- a/compiler/rustc_codegen_ssa/src/back/link.rs
+++ b/compiler/rustc_codegen_ssa/src/back/link.rs
@@ -58,6 +58,7 @@ use super::linker::{self, Linker};
 use super::metadata::{MetadataPosition, create_wrapper_file};
 use super::rpath::{self, RPathConfig};
 use super::{apple, versioned_llvm_target};
+use crate::base::needs_allocator_shim_for_linking;
 use crate::{
     CodegenResults, CompiledModule, CrateInfo, NativeLib, errors, looks_like_rust_object_file,
 };
@@ -2080,9 +2081,17 @@ fn add_local_crate_regular_objects(cmd: &mut dyn Linker, codegen_results: &Codeg
 }
 
 /// Add object files for allocator code linked once for the whole crate tree.
-fn add_local_crate_allocator_objects(cmd: &mut dyn Linker, codegen_results: &CodegenResults) {
-    if let Some(obj) = codegen_results.allocator_module.as_ref().and_then(|m| m.object.as_ref()) {
-        cmd.add_object(obj);
+fn add_local_crate_allocator_objects(
+    cmd: &mut dyn Linker,
+    codegen_results: &CodegenResults,
+    crate_type: CrateType,
+) {
+    if needs_allocator_shim_for_linking(&codegen_results.crate_info.dependency_formats, crate_type)
+    {
+        if let Some(obj) = codegen_results.allocator_module.as_ref().and_then(|m| m.object.as_ref())
+        {
+            cmd.add_object(obj);
+        }
     }
 }
 
@@ -2281,7 +2290,7 @@ fn linker_with_args(
         codegen_results,
         metadata,
     );
-    add_local_crate_allocator_objects(cmd, codegen_results);
+    add_local_crate_allocator_objects(cmd, codegen_results, crate_type);
 
     // Avoid linking to dynamic libraries unless they satisfy some undefined symbols
     // at the point at which they are specified on the command line.

--- a/compiler/rustc_codegen_ssa/src/back/linker.rs
+++ b/compiler/rustc_codegen_ssa/src/back/linker.rs
@@ -11,8 +11,9 @@ use rustc_metadata::{
 };
 use rustc_middle::bug;
 use rustc_middle::middle::dependency_format::Linkage;
-use rustc_middle::middle::exported_symbols;
-use rustc_middle::middle::exported_symbols::{ExportedSymbol, SymbolExportInfo, SymbolExportKind};
+use rustc_middle::middle::exported_symbols::{
+    self, ExportedSymbol, SymbolExportInfo, SymbolExportKind, SymbolExportLevel,
+};
 use rustc_middle::ty::TyCtxt;
 use rustc_session::Session;
 use rustc_session::config::{self, CrateType, DebugInfo, LinkerPluginLto, Lto, OptLevel, Strip};
@@ -22,6 +23,8 @@ use tracing::{debug, warn};
 
 use super::command::Command;
 use super::symbol_export;
+use crate::back::symbol_export::allocator_shim_symbols;
+use crate::base::needs_allocator_shim_for_linking;
 use crate::errors;
 
 #[cfg(test)]
@@ -1837,6 +1840,14 @@ fn exported_symbols_for_non_proc_macro(
             symbol_export::extend_exported_symbols(&mut symbols, tcx, symbol, cnum);
         }
     });
+
+    // Mark allocator shim symbols as exported only if they were generated.
+    if export_threshold == SymbolExportLevel::Rust
+        && needs_allocator_shim_for_linking(tcx.dependency_formats(()), crate_type)
+        && tcx.allocator_kind(()).is_some()
+    {
+        symbols.extend(allocator_shim_symbols(tcx));
+    }
 
     symbols
 }

--- a/compiler/rustc_codegen_ssa/src/back/linker.rs
+++ b/compiler/rustc_codegen_ssa/src/back/linker.rs
@@ -1827,7 +1827,7 @@ fn exported_symbols_for_non_proc_macro(
     let export_threshold = symbol_export::crates_export_threshold(&[crate_type]);
     for_each_exported_symbols_include_dep(tcx, crate_type, |symbol, info, cnum| {
         // Do not export mangled symbols from cdylibs and don't attempt to export compiler-builtins
-        // from any cdylib. The latter doesn't work anyway as we use hidden visibility for
+        // from any dylib. The latter doesn't work anyway as we use hidden visibility for
         // compiler-builtins. Most linkers silently ignore it, but ld64 gives a warning.
         if info.level.is_below_threshold(export_threshold) && !tcx.is_compiler_builtins(cnum) {
             symbols.push((

--- a/compiler/rustc_codegen_ssa/src/back/lto.rs
+++ b/compiler/rustc_codegen_ssa/src/back/lto.rs
@@ -8,8 +8,9 @@ use rustc_middle::ty::TyCtxt;
 use rustc_session::config::{CrateType, Lto};
 use tracing::info;
 
-use crate::back::symbol_export::{self, symbol_name_for_instance_in_crate};
+use crate::back::symbol_export::{self, allocator_shim_symbols, symbol_name_for_instance_in_crate};
 use crate::back::write::CodegenContext;
+use crate::base::allocator_kind_for_codegen;
 use crate::errors::{DynamicLinkingWithLTO, LtoDisallowed, LtoDylib, LtoProcMacro};
 use crate::traits::*;
 
@@ -113,6 +114,11 @@ pub(super) fn exported_symbols_for_lto(
             let _timer = tcx.prof.generic_activity("lto_generate_symbols_below_threshold");
             symbols_below_threshold.extend(copy_symbols(cnum));
         }
+    }
+
+    // Mark allocator shim symbols as exported only if they were generated.
+    if export_threshold == SymbolExportLevel::Rust && allocator_kind_for_codegen(tcx).is_some() {
+        symbols_below_threshold.extend(allocator_shim_symbols(tcx).map(|(name, _kind)| name));
     }
 
     symbols_below_threshold

--- a/compiler/rustc_codegen_ssa/src/back/symbol_export.rs
+++ b/compiler/rustc_codegen_ssa/src/back/symbol_export.rs
@@ -18,7 +18,7 @@ use rustc_symbol_mangling::mangle_internal_symbol;
 use rustc_target::spec::TlsModel;
 use tracing::debug;
 
-use crate::base::allocator_kind_for_codegen;
+use crate::back::symbol_export;
 
 fn threshold(tcx: TyCtxt<'_>) -> SymbolExportLevel {
     crates_export_threshold(tcx.crate_types())
@@ -215,31 +215,6 @@ fn exported_non_generic_symbols_provider_local<'tcx>(
                 rustc_std_internal_symbol: false,
             },
         ));
-    }
-
-    // Mark allocator shim symbols as exported only if they were generated.
-    if allocator_kind_for_codegen(tcx).is_some() {
-        for symbol_name in ALLOCATOR_METHODS
-            .iter()
-            .map(|method| mangle_internal_symbol(tcx, global_fn_name(method.name).as_str()))
-            .chain([
-                mangle_internal_symbol(tcx, "__rust_alloc_error_handler"),
-                mangle_internal_symbol(tcx, OomStrategy::SYMBOL),
-                mangle_internal_symbol(tcx, NO_ALLOC_SHIM_IS_UNSTABLE),
-            ])
-        {
-            let exported_symbol = ExportedSymbol::NoDefId(SymbolName::new(tcx, &symbol_name));
-
-            symbols.push((
-                exported_symbol,
-                SymbolExportInfo {
-                    level: SymbolExportLevel::Rust,
-                    kind: SymbolExportKind::Text,
-                    used: false,
-                    rustc_std_internal_symbol: true,
-                },
-            ));
-        }
     }
 
     // Sort so we get a stable incr. comp. hash.
@@ -514,6 +489,31 @@ pub(crate) fn provide(providers: &mut Providers) {
     providers.extern_queries.is_reachable_non_generic = is_reachable_non_generic_provider_extern;
     providers.extern_queries.upstream_monomorphizations_for =
         upstream_monomorphizations_for_provider;
+}
+
+pub(crate) fn allocator_shim_symbols(
+    tcx: TyCtxt<'_>,
+) -> impl Iterator<Item = (String, SymbolExportKind)> {
+    ALLOCATOR_METHODS
+        .iter()
+        .map(move |method| mangle_internal_symbol(tcx, global_fn_name(method.name).as_str()))
+        .chain([
+            mangle_internal_symbol(tcx, "__rust_alloc_error_handler"),
+            mangle_internal_symbol(tcx, OomStrategy::SYMBOL),
+            mangle_internal_symbol(tcx, NO_ALLOC_SHIM_IS_UNSTABLE),
+        ])
+        .map(move |symbol_name| {
+            let exported_symbol = ExportedSymbol::NoDefId(SymbolName::new(tcx, &symbol_name));
+
+            (
+                symbol_export::exporting_symbol_name_for_instance_in_crate(
+                    tcx,
+                    exported_symbol,
+                    LOCAL_CRATE,
+                ),
+                SymbolExportKind::Text,
+            )
+        })
 }
 
 fn symbol_export_level(tcx: TyCtxt<'_>, sym_def_id: DefId) -> SymbolExportLevel {

--- a/compiler/rustc_codegen_ssa/src/base.rs
+++ b/compiler/rustc_codegen_ssa/src/base.rs
@@ -17,6 +17,7 @@ use rustc_hir::lang_items::LangItem;
 use rustc_hir::{ItemId, Target};
 use rustc_middle::middle::codegen_fn_attrs::CodegenFnAttrs;
 use rustc_middle::middle::debugger_visualizer::{DebuggerVisualizerFile, DebuggerVisualizerType};
+use rustc_middle::middle::dependency_format::Dependencies;
 use rustc_middle::middle::exported_symbols::{self, SymbolExportKind};
 use rustc_middle::middle::lang_items;
 use rustc_middle::mir::BinOp;
@@ -630,14 +631,30 @@ pub fn allocator_kind_for_codegen(tcx: TyCtxt<'_>) -> Option<AllocatorKind> {
     // If the crate doesn't have an `allocator_kind` set then there's definitely
     // no shim to generate. Otherwise we also check our dependency graph for all
     // our output crate types. If anything there looks like its a `Dynamic`
-    // linkage, then it's already got an allocator shim and we'll be using that
-    // one instead. If nothing exists then it's our job to generate the
-    // allocator!
-    let any_dynamic_crate = tcx.dependency_formats(()).iter().any(|(_, list)| {
+    // linkage for all crate types we may link as, then it's already got an
+    // allocator shim and we'll be using that one instead. If nothing exists
+    // then it's our job to generate the allocator! If crate types disagree
+    // about whether an allocator shim is necessary or not, we generate one
+    // and let needs_allocator_shim_for_linking decide at link time whether or
+    // not to use it for any particular linker invocation.
+    let all_crate_types_any_dynamic_crate = tcx.dependency_formats(()).iter().all(|(_, list)| {
         use rustc_middle::middle::dependency_format::Linkage;
         list.iter().any(|&linkage| linkage == Linkage::Dynamic)
     });
-    if any_dynamic_crate { None } else { tcx.allocator_kind(()) }
+    if all_crate_types_any_dynamic_crate { None } else { tcx.allocator_kind(()) }
+}
+
+/// Decide if this particular crate type needs an allocator shim linked in.
+/// This may return true even when allocator_kind_for_codegen returns false. In
+/// this case no allocator shim shall be linked.
+pub(crate) fn needs_allocator_shim_for_linking(
+    dependency_formats: &Dependencies,
+    crate_type: CrateType,
+) -> bool {
+    use rustc_middle::middle::dependency_format::Linkage;
+    let any_dynamic_crate =
+        dependency_formats[&crate_type].iter().any(|&linkage| linkage == Linkage::Dynamic);
+    !any_dynamic_crate
 }
 
 pub fn codegen_crate<B: ExtraBackendMethods>(

--- a/tests/ui/linking/mixed-allocator-shim.rs
+++ b/tests/ui/linking/mixed-allocator-shim.rs
@@ -1,0 +1,15 @@
+//@ build-pass
+//@ compile-flags: --crate-type staticlib,dylib -Zstaticlib-prefer-dynamic
+//@ no-prefer-dynamic
+
+// Test that compiling for multiple crate types in a single compilation with
+// mismatching allocator shim requirements doesn't result in the allocator shim
+// missing entirely.
+// In this particular test the dylib crate type will statically link libstd and
+// thus need an allocator shim, while the staticlib crate type will dynamically
+// link libstd and thus not need an allocator shim.
+// The -Zstaticlib-prefer-dynamic flag could be avoided by doing it the other
+// way around, but testing that the staticlib correctly has the allocator shim
+// in that case would require a run-make test instead.
+
+pub fn foo() {}

